### PR TITLE
Add info log message about duration taken to load plugins

### PR DESCRIPTION
**Pull request by ashb**
Loading plugins, particularly from setuptools entry points can be slow,
and since by default this happens per-task, it can slow down task
execution unexpectedly.

By having this log message users can know the source of the delay

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Review by kaxil — APPROVED** [review]

**Review by mik-laj — APPROVED** [review]

File: airflow/utils/log/colored_log.py [hunk 1]
```diff
@@ -90,8 +90,9 @@ def _color_record_traceback(self, record: LogRecord) -> LogRecord:
 
     def format(self, record: LogRecord) -> str:
         try:
-            record = self._color_record_args(record)
-            record = self._color_record_traceback(record)
+            if self.stream.isatty():
```
Comment (ashb) [inline airflow/utils/log/colored_log.py:93]: This is so that `caplog` fixture from pytest doesn't get poluted with escape code.

File: tests/plugins/test_plugins_manager.py [hunk 2]
```diff
@@ -163,14 +170,16 @@ class AirflowAdminMenuLinksPlugin(AirflowPlugin):
...
             appbuilder_menu_items = [mock.MagicMock()]
 
-        with mock_plugin_manager(plugins=[AirflowAdminViewsPlugin(), AirflowAdminMenuLinksPlugin()]):
+        with mock_plugin_manager(
+            plugins=[AirflowAdminViewsPlugin(), AirflowAdminMenuLinksPlugin()]
+        ), caplog.at_level(logging.WARNING):
             from airflow import plugins_manager
 
-            # assert not logs
-            with self.assertRaises(AssertionError), self.assertLogs(plugins_manager.log):
```
Comment (ashb) [inline tests/plugins/test_plugins_manager.py:170]: I found this a huge code smell, and a funny/hard to follow way of saying "does not log".

So I switched it all out to pytests `caplog` fixture.

File: tests/plugins/test_plugins_manager.py [hunk 3]
```diff
@@ -119,10 +120,13 @@ class TestPropertyHook(BaseHook):
...
             plugins_manager.ensure_plugins_loaded()
 
-            self.assertIn('AirflowTestPropertyPlugin', str(plugins_manager.plugins))
-            self.assertIn("TestPropertyHook", str(plugins_manager.registered_hooks))
+            assert 'AirflowTestPropertyPlugin' in str(plugins_manager.plugins)
+            assert 'TestPropertyHook' in str(plugins_manager.registered_hooks)
 
-    def test_should_warning_about_incompatible_plugins(self):
+        assert caplog.records[0].levelname == 'INFO'
+        assert caplog.records[0].msg == 'Loading %d plugin(s) took %.2f seconds'
```
Comment (ashb) [inline tests/plugins/test_plugins_manager.py:136]: his is the un-formatted message string

File: airflow/plugins_manager.py [hunk 4]
```diff
@@ -262,6 +263,8 @@ def ensure_plugins_loaded():
 
     log.debug("Loading plugins")
 
+    start = time.monotonic()
```
Comment (mjpieters) [inline airflow/plugins_manager.py:266]: Might there be a usecase for a general timing context manager here?

```python
with timer() as t:
    # ...

log.info("Frobbing the foonars took %.2f seconds", t.elapsed.total_seconds())
```

Encapsulation makes it easier to make sure everyone gets this right (using `time.monotonic()`!), and might make it easier to automatically export metrics to statd / prometheus if desired (`with timer(metric="metric_name") as t:`).

Comment (ashb) [reply airflow/plugins_manager.py:266]: Could be. For metrics there is already a `with Stats.timer("my_metric_name":)` though.

Comment (ashb) [reply airflow/plugins_manager.py:266]: #12353

Comment (turbaszek) [reply airflow/plugins_manager.py:266]: Only problem I see is that context manager introduces additional level of indentation which can require a pylint disable sometimes... Apart from that I like the idea!

File: airflow/plugins_manager.py [hunk 5]
```diff
@@ -273,6 +276,12 @@ def ensure_plugins_loaded():
     for plugin in plugins:
         registered_hooks.extend(plugin.hooks)
 
+    end = time.monotonic()
+
+    num_loaded = len(plugins)
+    if num_loaded > 0:
+        log.info("Loading %d plugin(s) took %.2f seconds", num_loaded, end - start)
```
Comment (turbaszek) [inline airflow/plugins_manager.py:283]:
```suggestion
    if plugins:
        log.info("Loading %d plugin(s) took %.2f seconds", len(plugins), end - start)
```
WDYT?

**Review by turbaszek — APPROVED** [review]

File: airflow/plugins_manager.py (hunk 4 above)
Comment (ashb) [reply airflow/plugins_manager.py:266]: Can also do it like this:

```python
timer = Stats.timer().start()
...
timer.end()
# use timer.duration
```

**Comment by kaxil** [comment]
https://github.com/apache/airflow/pull/12308/checks?check_run_id=1399669598#step:6:1249

```
FAILED tests/plugins/test_plugins_manager.py::TestPluginsManager::test_should_load_plugins_from_property
FAILED tests/plugins/test_plugins_manager.py::TestPluginsManager::test_should_warning_about_incompatible_plugins
======= 2 failed, 368 passed, 8 skipped, 2 warnings in 203.14s (0:03:23) =======
```


```
=================================== FAILURES ===================================
__________ TestPluginsManager.test_should_load_plugins_from_property ___________

self = <tests.plugins.test_plugins_manager.TestPluginsManager object at 0x7efdc0b1c890>
caplog = <_pytest.logging.LogCaptureFixture object at 0x7efdc0a27d50>

    def test_should_load_plugins_from_property(self, caplog):
        class AirflowTestPropertyPlugin(AirflowPlugin):
            name = "test_property_plugin"
    
            @property
            def hooks(self):
                class TestPropertyHook(BaseHook):
                    pass
    
                return [TestPropertyHook]
    
        with mock_plugin_manager(plugins=[AirflowTestPropertyPlugin()]):
            from airflow import plugins_manager
    
            plugins_manager.ensure_plugins_loaded()
    
            assert 'AirflowTestPropertyPlugin' in str(plugins_manager.plugins)
            assert 'TestPropertyHook' in str(plugins_manager.registered_hooks)
    
>       assert caplog.records[0].levelname == 'INFO'
E       IndexError: list index out of range

tests/plugins/test_plugins_manager.py:135: IndexError
----------------------------- Captured stdout call -----------------------------
[2020-11-14 11:18:08,276] {{plugins_manager.py:282}} INFO - Loading 1 plugin(s) took 0.00 seconds
______ TestPluginsManager.test_should_warning_about_incompatible_plugins _______

self = <tests.plugins.test_plugins_manager.TestPluginsManager object at 0x7efdc0a8b310>
caplog = <_pytest.logging.LogCaptureFixture object at 0x7efdc0a8b910>

    def test_should_warning_about_incompatible_plugins(self, caplog):
        class AirflowAdminViewsPlugin(AirflowPlugin):
            name = "test_admin_views_plugin"
    
            admin_views = [mock.MagicMock()]
    
        class AirflowAdminMenuLinksPlugin(AirflowPlugin):
            name = "test_menu_links_plugin"
    
            menu_links = [mock.MagicMock()]
    
        with mock_plugin_manager(
            plugins=[AirflowAdminViewsPlugin(), AirflowAdminMenuLinksPlugin()]
        ), caplog.at_level(logging.WARNING):
            from airflow import plugins_manager
    
            plugins_manager.initialize_web_ui_plugins()
    
>       assert caplog.record_tuples == [
            (
                "airflow.plugins_manager",
                logging.WARNING,
                "Plugin 'test_admin_views_plugin' may not be compatible with the current Airflow version. "
                "Please contact the author of the plugin.",
            ),
            (
                "airflow.plugins_manager",
                logging.WARNING,
                "Plugin 'test_menu_links_plugin' may not be compatible with the current Airflow version. "
                "Please contact the author of the plugin.",
            ),
        ]
E       assert [] == [('airflow.pl...the plugin.")]
E         Right contains 2 more items, first extra item: ('airflow.plugins_manager', 30, "Plugin 'test_admin_views_plugin' may not be compatible with the current Airflow version. Please contact the author of the plugin.")
E         Use -v to get the full diff

tests/plugins/test_plugins_manager.py:156: AssertionError
```

**Comment by ashb** [comment]
Hmmmmm, passes for me locally. Fun.
